In [42]:
import pandas as pd
import numpy as np #storing as array
from sklearn import tree
from sklearn import preprocessing 
# used for finding missing values, text into numrical

In [43]:
train = pd.read_csv(r'C:\Users\mansi\Documents\LetsUpgrade\Day 24\Decision_Tree_Titanic_Survival\train.csv')
train.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [44]:
train.describe().T #Checking if any attribute ishaving standard deviation zero.

count        mean         std   min       25%       50%    75%  \
PassengerId  889.0  446.000000  256.998173  1.00  224.0000  446.0000  668.0   
Survived     889.0    0.382452    0.486260  0.00    0.0000    0.0000    1.0   
Pclass       889.0    2.311586    0.834700  1.00    2.0000    3.0000    3.0   
Age          889.0   32.699854   14.346821  0.42   22.0000   32.0000   45.0   
SibSp        889.0    0.524184    1.103705  0.00    0.0000    0.0000    1.0   
Parch        889.0    0.382452    0.806761  0.00    0.0000    0.0000    0.0   
Fare         889.0   32.096681   49.697504  0.00    7.8958   14.4542   31.0   

                  max  
PassengerId  891.0000  
Survived       1.0000  
Pclass         3.0000  
Age           80.0000  
SibSp          8.0000  
Parch          6.0000  
Fare         512.3292

In [45]:
train.shape

(889, 12)

In [46]:
#finding missing values
train.isna().sum() #Checking NA/Null values  #cabin having NA values.

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [47]:
train["Age"].mean()

32.69985376827896

In [48]:
new_age=np.where(train["Age"].isnull(),32,train["Age"])

train["Age"]=new_age

train['Sex'] = train['Sex'].map({'male':1, 'female': 0})

train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1      0   
4                           Allen, Mr. William Henry    1  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

In [49]:
data1=pd.DataFrame([train["Sex"],train["Age"],train["Fare"]]).T


In [50]:

treemodel=tree.DecisionTreeClassifier(max_depth=6)



In [51]:
treemodel.fit(X=data1,y=train["Survived"])

DecisionTreeClassifier(max_depth=6)

In [52]:
treemodel.score(X=data1,y=train["Survived"])

0.829021372328459

In [53]:
# Prediction of Survival Using Age,Sex,Fare

test=pd.read_csv(r"C:\Users\mansi\Documents\LetsUpgrade\Day 24\Decision_Tree_Titanic_Survival\test.csv")

test["Age"].mean()

33.23741007194245

new_age2=np.where(test["Age"].isnull(),33,test["Age"])

test["Age"]=new_age2
test['Sex'] = test['Sex'].map({'male':1, 'female': 0})

test.head()

PassengerId  Pclass                                          Name  Sex  \
0          892       3                              Kelly, Mr. James    1   
1          893       3              Wilkes, Mrs. James (Ellen Needs)    0   
2          894       2                     Myles, Mr. Thomas Francis    1   
3          895       3                              Wirz, Mr. Albert    1   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)    0   

    Age  SibSp  Parch   Ticket     Fare Embarked  
0  34.5      0      0   330911   7.8292        Q  
1  47.0      1      0   363272   7.0000        S  
2  62.0      0      0   240276   9.6875        Q  
3  27.0      0      0   315154   8.6625        S  
4  22.0      1      1  3101298  12.2875        S

In [54]:
data2=pd.DataFrame([test["Sex"], test["Age"],test["Fare"]]).T

test_preds=treemodel.predict(X=data2)

predicted_output=pd.DataFrame({"PassengerId":test["PassengerId"],"Survived":test_preds})

predicted_output.to_csv("output.csv",index=False);

 ## Conclusion:
    
 Out of 889 people , 312 is female and 577 is male .

 38 female whose fare is lesser than 10.481 have survived.
    
 70 female  whose fare is between 10.481 and 24.808 have survived.

 14 female whose fare is more than 25.698 and age is more than 27 have survived.
    
 6 female whose fare is between 36.88 and 44.24 have survived.

 15 female whose age is below 24.5 and fare is between 69.425 have survived.
    
 5 male whose age is below 3.5 and fare is lesser than 66.379 have survived.

 1 male whose fare is between 17.344 and 19.656 and age is lesser than 13.5 have survived.
    
 20 male who were below 30.75 and whose fare is lesser than 26.269 have survived.

 12 male whose fare is between 7.91 and 26.269 and age below 32.5 have survived.
    
 3 male whose age is below 22.5 and fare lesser than 387.665 have survived.

 40 male whose age greater than 22.5 and fare is 387.665 have survived. 



In [55]:
with open('DecisionTree1.dot','w') as file:    
    file = tree.export_graphviz(treemodel,feature_names=['Sex','Age','Fare'], out_file= file)